AssertionError in rule make_contigs_kmer_index

```
rule make_contigs_kmer_index:
    input: rgnv_nbhd_k31/bcalm.unitigs.db
    output: rgnv_nbhd_k31_r5/contigs.mphf, rgnv_nbhd_k31_r5/contigs.indices
    jobid: 2


        /home/tereiter/miniconda3/envs/sgc_dom_abund/bin/python -m spacegraphcats.cdbg.index_cdbg_by_kmer             -k 31 rg
nv_nbhd_k31_r5 --contigs-db rgnv_nbhd_k31/bcalm.unitigs.db

/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/runpy.py:127: RuntimeWarning: 'spacegraphcats.cdbg.index_cdbg_by_km
er' found in sys.modules after import of package 'spacegraphcats.cdbg', but prior to execution of 'spacegraphcats.cdbg.index_c
dbg_by_kmer'; this may result in unpredictable behaviour
  warn(RuntimeWarning(msg))
reading cDBG nodes from sqlite DB rgnv_nbhd_k31/bcalm.unitigs.db
loaded 46667471 contigs.

building MPHF for 519194226 k-mers in 46667471 nodes.
second pass.
reading cDBG nodes from sqlite DB rgnv_nbhd_k31/bcalm.unitigs.db
loaded 46667471 contigs in pass2.

Traceback (most recent call last):
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/site-packages/spacegraphcats/cdbg/index_cdbg_by_kmer.py", l
ine 216, in <module>
    sys.exit(main(sys.argv[1:]))
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/site-packages/spacegraphcats/cdbg/index_cdbg_by_kmer.py", l
ine 205, in main
    table, sizes = build_mphf(a.ksize, create_records_iter)
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/site-packages/spacegraphcats/cdbg/index_cdbg_by_kmer.py", l
ine 181, in build_mphf
    assert n == max(table.mphf_to_value), (n, max(table.mphf_to_value))
AssertionError: (46667470, 4294967295)
```

When I've had errors like this in the past, some earlier sgc steps have silently failed due to insufficient RAM, but snakemake has continued running. I can't assess whether that is the problem here because i ran this in an srun session and can't scroll back that far. sigh. 


I installed sgc with:
```
conda env create -f sgc_env.yml -n sgc_dom_abund
conda activate sgc_dominator_abund
pip install git+https://github.com/spacegraphcats/spacegraphcats@dominator_abund
```

where `sgc_env.yml` was

```
name: sgc_dom_abund
channels:
    - conda-forge
    - bioconda
    - defaults
dependencies:
    - bcalm=2.2.3
    - snakemake-minimal=5.20.1
    - screed=1.0.1
    - khmer=3.0.0a3
    - numpy
    - pytest=5.3.2
    - sortedcontainers=2.1.0
    - click=7.1.2
    - pyflakes
    - bbhash>=0.5.4
    - pip
    - pip:
      - sourmash==4.0.0a3
```

I'm also attaching my conf file:
[rgnv_nbhd_catlas_r5_conf.yml.txt](https://github.com/spacegraphcats/spacegraphcats/files/5667322/rgnv_nbhd_catlas_r5_conf.yml.txt)

Same exact error message, this time with `This is spacegraphcats 2.0b11.`

```
rule make_contigs_kmer_index:
    input: rgnv_nbhd_k31/bcalm.unitigs.db
    output: rgnv_nbhd_k31_r5/contigs.mphf, rgnv_nbhd_k31_r5/contigs.indices
    jobid: 2


This is spacegraphcats 2.0b11.

Job counts:
        count   jobs
        1       make_contigs_kmer_index
        1
/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/runpy.py:127: RuntimeWarning: 'spacegraphcats.cdbg.index_cdbg_by_kmer' foun$
 in sys.modules after import of package 'spacegraphcats.cdbg', but prior to execution of 'spacegraphcats.cdbg.index_cdbg_by_kmer'; thi$
 may result in unpredictable behaviour
  warn(RuntimeWarning(msg))
reading cDBG nodes from sqlite DB rgnv_nbhd_k31/bcalm.unitigs.db
loaded 46667471 contigs.

building MPHF for 519194226 k-mers in 46667471 nodes.
second pass.
reading cDBG nodes from sqlite DB rgnv_nbhd_k31/bcalm.unitigs.db
loaded 46667471 contigs in pass2.

Traceback (most recent call last):
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/site-packages/spacegraphcats/cdbg/index_cdbg_by_kmer.py", line 216,
in <module>
    sys.exit(main(sys.argv[1:]))
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/site-packages/spacegraphcats/cdbg/index_cdbg_by_kmer.py", line 205,
in main
    table, sizes = build_mphf(a.ksize, create_records_iter)
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/site-packages/spacegraphcats/cdbg/index_cdbg_by_kmer.py", line 181,
in build_mphf
    assert n == max(table.mphf_to_value), (n, max(table.mphf_to_value))
AssertionError: (46667470, 4294967295)
```

--- a/spacegraphcats/cdbg/index_cdbg_by_kmer.py
+++ b/spacegraphcats/cdbg/index_cdbg_by_kmer.py
@@ -139,6 +139,7 @@ class MPHF_KmerIndex(object):
 def build_mphf(ksize, records_iter_fn):
     # build a list of all k-mers in the cDBG
     all_kmers = set()
+    sum_kmers = 0
 
     records_iter = records_iter_fn()
     for n, record in enumerate(records_iter):
@@ -146,10 +147,16 @@ def build_mphf(ksize, records_iter_fn):
             print("... contig", n, end="\r")
 
         kmers = hash_sequence(record.sequence, ksize)
+        sum_kmers += len(kmers)
         all_kmers.update(kmers)
 
     n_contigs = n + 1
     print(f"loaded {n_contigs} contigs.\n")
+
+    if len(all_kmers) != sum_kmers:
+        print('WARNING: likely hash collisions (or duplicate k-mers?) in input cDBG')
+        print('WARNING: we hashed {sum_kmers}, but only {len(all_kmers)} distinct hashes.')
+        print('WARNING: the impact of this on spacegraphcats is unclear, but, at least for now, there\'s nothing you can do about it. Apologies.')
 
     # build MPHF (this is the CPU intensive bit)
     print(f"building MPHF for {len(all_kmers)} k-mers in {n_contigs} nodes.")

--- a/spacegraphcats/cdbg/index_cdbg_by_kmer.py
+++ b/spacegraphcats/cdbg/index_cdbg_by_kmer.py
@@ -16,7 +16,6 @@ import os
 import argparse
 import pickle
 import sqlite3
-import collections
 
 import screed
 import khmer
@@ -139,8 +138,9 @@ class MPHF_KmerIndex(object):
 
 def build_mphf(ksize, records_iter_fn):
     # build a list of all k-mers in the cDBG
-    all_kmers = collections.Counter()
+    all_kmers = set()
     sum_kmers = 0
+    multicounts = set()
 
     records_iter = records_iter_fn()
     for n, record in enumerate(records_iter):
@@ -149,31 +149,30 @@ def build_mphf(ksize, records_iter_fn):
 
         kmers = hash_sequence(record.sequence, ksize)
         sum_kmers += len(kmers)
-        all_kmers.update(kmers)
+        these_kmers = set(kmers)
+
+        # have we seen any of these kmers before? if so, remove.
+        seen_before = all_kmers.intersection(these_kmers)
+        if seen_before:
+            multicounts.update(seen_before)
+            these_kmers -= seen_before
+
+        all_kmers.update(these_kmers)
 
     n_contigs = n + 1
     print(f"loaded {n_contigs} contigs.\n")
-
-    multicounts = set()
-    for (hashval, count) in all_kmers.most_common():
-        if count == 1:
-            break
-        multicounts.add(hashval)
 
     if multicounts:
         print('NOTE: likely hash collisions (or duplicate k-mers?) in input cDBG')
         print(f'NOTE: {len(multicounts)} k-mer hash values are present more than once.')
         print('NOTE: these k-mers are being removed from consideration.')
-
-        for hashval in multicounts:
-            del all_kmers[hashval]
     else:
         print('NOTE: no multicount hashvals detected.')
 
     # build MPHF (this is the CPU intensive bit)
     print(f"building MPHF for {len(all_kmers)} k-mers in {n_contigs} nodes.")
     table = BBHashTable()
-    table.initialize(list(all_kmers.keys()))
+    table.initialize(list(all_kmers))
 
     del all_kmers
 
@@ -197,9 +196,8 @@ def build_mphf(ksize, records_iter_fn):
         kmers = hash_sequence(record.sequence, ksize)
 
         # for each k-mer, find its MPHF hashval, & link to info.
-        for kmer in kmers:
-            if kmer not in multicounts:
-                table[kmer] = cdbg_id
+        for kmer in set(kmers) - multicounts:
+            table[kmer] = cdbg_id
 
         sizes[cdbg_id] = len(kmers)
 

--- a/spacegraphcats/cdbg/index_cdbg_by_kmer.py
+++ b/spacegraphcats/cdbg/index_cdbg_by_kmer.py
@@ -155,7 +155,6 @@ def build_mphf(ksize, records_iter_fn):
         seen_before = all_kmers.intersection(these_kmers)
         if seen_before:
             multicounts.update(seen_before)
-            these_kmers -= seen_before
 
         all_kmers.update(these_kmers)
 
@@ -166,6 +165,8 @@ def build_mphf(ksize, records_iter_fn):
         print('NOTE: likely hash collisions (or duplicate k-mers?) in input cDBG')
         print(f'NOTE: {len(multicounts)} k-mer hash values are present more than once.')
         print('NOTE: these k-mers are being removed from consideration.')
+
+        all_kmers -= multicounts
     else:
         print('NOTE: no multicount hashvals detected.')
 

--- a/spacegraphcats/cdbg/index_cdbg_by_kmer.py
+++ b/spacegraphcats/cdbg/index_cdbg_by_kmer.py
@@ -16,12 +16,13 @@ import os
 import argparse
 import pickle
 import sqlite3
+import collections
 
 import screed
 import khmer
 from bbhash_table import BBHashTable
 
-UNSET_VALUE=2**32-1             # BBHashTable "unset" value
+UNSET_VALUE = 2**32 - 1             # BBHashTable "unset" value
 
 
 hashing_fn = None
@@ -138,7 +139,7 @@ class MPHF_KmerIndex(object):
 
 def build_mphf(ksize, records_iter_fn):
     # build a list of all k-mers in the cDBG
-    all_kmers = set()
+    all_kmers = collections.Counter()
     sum_kmers = 0
 
     records_iter = records_iter_fn()
@@ -153,15 +154,26 @@ def build_mphf(ksize, records_iter_fn):
     n_contigs = n + 1
     print(f"loaded {n_contigs} contigs.\n")
 
-    if len(all_kmers) != sum_kmers:
-        print('WARNING: likely hash collisions (or duplicate k-mers?) in input cDBG')
-        print(f'WARNING: we hashed {sum_kmers}, but only {len(all_kmers)} distinct hashes.')
-        print('WARNING: the impact of this on spacegraphcats is unclear, but, at least for now, there\'s nothing you can do about it. Apologies.')
+    multicounts = set()
+    for (hashval, count) in all_kmers.most_common():
+        if count == 1:
+            break
+        multicounts.add(hashval)
+
+    if multicounts:
+        print('NOTE: likely hash collisions (or duplicate k-mers?) in input cDBG')
+        print(f'NOTE: {len(multicounts)} k-mer hash values are present more than once.')
+        print('NOTE: these k-mers are being removed from consideration.')
+
+        for hashval in multicounts:
+            del all_kmers[hashval]
+    else:
+        print('NOTE: no multicount hashvals detected.')
 
     # build MPHF (this is the CPU intensive bit)
     print(f"building MPHF for {len(all_kmers)} k-mers in {n_contigs} nodes.")
     table = BBHashTable()
-    table.initialize(list(all_kmers))
+    table.initialize(list(all_kmers.keys()))
 
     del all_kmers
 
@@ -186,7 +198,8 @@ def build_mphf(ksize, records_iter_fn):
 
         # for each k-mer, find its MPHF hashval, & link to info.
         for kmer in kmers:
-            table[kmer] = cdbg_id
+            if kmer not in multicounts:
+                table[kmer] = cdbg_id
 
         sizes[cdbg_id] = len(kmers)
 

--- a/spacegraphcats/cdbg/index_cdbg_by_kmer.py
+++ b/spacegraphcats/cdbg/index_cdbg_by_kmer.py
@@ -155,7 +155,7 @@ def build_mphf(ksize, records_iter_fn):
 
     if len(all_kmers) != sum_kmers:
         print('WARNING: likely hash collisions (or duplicate k-mers?) in input cDBG')
-        print('WARNING: we hashed {sum_kmers}, but only {len(all_kmers)} distinct hashes.')
+        print(f'WARNING: we hashed {sum_kmers}, but only {len(all_kmers)} distinct hashes.')
         print('WARNING: the impact of this on spacegraphcats is unclear, but, at least for now, there\'s nothing you can do about it. Apologies.')
 
     # build MPHF (this is the CPU intensive bit)

--- a/spacegraphcats/cdbg/index_cdbg_by_kmer.py
+++ b/spacegraphcats/cdbg/index_cdbg_by_kmer.py
@@ -21,6 +21,8 @@ import screed
 import khmer
 from bbhash_table import BBHashTable
 
+UNSET_VALUE=2**32-1             # BBHashTable "unset" value
+
 
 hashing_fn = None
 hashing_ksize = None
@@ -136,7 +138,7 @@ class MPHF_KmerIndex(object):
 
 def build_mphf(ksize, records_iter_fn):
     # build a list of all k-mers in the cDBG
-    all_kmers = list()
+    all_kmers = set()
 
     records_iter = records_iter_fn()
     for n, record in enumerate(records_iter):
@@ -144,7 +146,7 @@ def build_mphf(ksize, records_iter_fn):
             print("... contig", n, end="\r")
 
         kmers = hash_sequence(record.sequence, ksize)
-        all_kmers.extend(list(kmers))
+        all_kmers.update(kmers)
 
     n_contigs = n + 1
     print(f"loaded {n_contigs} contigs.\n")
@@ -152,7 +154,9 @@ def build_mphf(ksize, records_iter_fn):
     # build MPHF (this is the CPU intensive bit)
     print(f"building MPHF for {len(all_kmers)} k-mers in {n_contigs} nodes.")
     table = BBHashTable()
-    table.initialize(all_kmers)
+    table.initialize(list(all_kmers))
+
+    del all_kmers
 
     # build tables linking:
     # * mphf hash to k-mer hash (for checking exactness)
@@ -162,12 +166,14 @@ def build_mphf(ksize, records_iter_fn):
     print("second pass.")
     records_iter = records_iter_fn()
     sizes = {}
+    max_cdbg_id = 0
     for n, record in enumerate(records_iter):
         if n % 50000 == 0 and n:
             print("... contig {} of {}".format(n, n_contigs), end="\r")
 
         # node ID is record name, must go from 0 to total-1
         cdbg_id = int(record.name)
+
         # get 64-bit numbers for each k-mer
         kmers = hash_sequence(record.sequence, ksize)
 
@@ -177,8 +183,27 @@ def build_mphf(ksize, records_iter_fn):
 
         sizes[cdbg_id] = len(kmers)
 
+        # update max cdbg_id:
+        if cdbg_id > max_cdbg_id:
+            max_cdbg_id = cdbg_id
+
     print(f"loaded {n_contigs} contigs in pass2.\n")
-    assert n == max(table.mphf_to_value), (n, max(table.mphf_to_value))
+
+    # a few asserts - these are a bit redundant with each other, but
+    # are here to aid in debugging.
+
+    # number of contigs should not be super large.
+    assert n <= UNSET_VALUE
+
+    # no values in the bbhash table should be unset.
+    max_table_value = max(table.mphf_to_value)
+    assert max_table_value != UNSET_VALUE
+
+    # cdbg_id should go from 0 to n_contigs - 1.
+    assert n == max_cdbg_id
+
+    # and max value in the bbhash table should be the max cdbg_id
+    assert n == max_table_value, (n, max_table_value)
 
     return table, sizes
 

